File: leetcode-go/Q051-Q100/Q068_Text_Justification/Q068_Text_Justification.go
package Q068_Text_Justification

func fullJustify(words []string, maxWidth int) []string {
	ans := []string{}
	begin := 0
	for begin < len(words) {
		last := begin
		count := len(words[last])
		last++
		for last < len(words) && count+len(words[last])+1 <= maxWidth {
			count += len(words[last]) + 1
			last++
		}

		space, extra := 1, 0
		if last < len(words) && last != begin+1 {
			space = (maxWidth-count)/(last-begin-1) + 1
			extra = (maxWidth - count) % (last - begin - 1)
		}

		line := []byte(words[begin])
		begin++

		for ; extra > 0; extra-- {
			line = appendnspace(line, space+1)
			line = append(line, words[begin]...)
			begin++
		}
		for begin < last {
			line = appendnspace(line, space)
			line = append(line, words[begin]...)
			begin++
		}
		line = appendnspace(line, maxWidth-len(line))

		ans = append(ans, string(line))
	}
	return ans
}

func appendnspace(s []byte, n int) []byte {
	for ; n > 0; n-- {
		s = append(s, ' ')
	}
	return s
}

File: src/Y2018/facebook/Q068_Text_Justification.java
package Y2018.facebook;

import java.util.LinkedList;
import java.util.List;

public class Q068_Text_Justification {

    public List<String> fullJustify(String[] words, int maxWidth) {
        List<String> result = new LinkedList<>();
        for (int i = 0; i < words.length; i++) {
            int l = i, r = i, width = 0;
            while (r < words.length && width <= maxWidth) {
                if (r != l) width += 1;
                width += words[r].length();
                r++;
            }
            r--;
            if (width > maxWidth) {
                r--;
            }

            width = 0;
            for (int j = l; j <= r; j++) {
                width += words[j].length();
            }
            int wordNum = r - l + 1;
            int spaceNum = wordNum == 1 ? 0 : (maxWidth - width) / (wordNum - 1);
            int extraSpace = wordNum == 1 ? maxWidth - width : (maxWidth - width) % (wordNum - 1);

            StringBuilder sb = new StringBuilder();
            if (r == words.length - 1) { // last line
                for (int j = l; j <= r; j++) {
                    if (j != l) sb.append(" ");
                    sb.append(words[j]);
                }
                while (sb.length() < maxWidth) {
                    sb.append(" ");
                }
            } else {
                for (int j = l; j <= r; j++) {
                    if (j != l) {
                        for (int k = 0; k < spaceNum; k++) {
                            sb.append(" ");
                        }
                    }
                    sb.append(words[j]);
                    if (l == r) {
                        while (extraSpace-- > 0) {
                            sb.append(" ");
                        }
                    } else {
                        if (extraSpace > 0) {
                            sb.append(" ");
                            extraSpace--;
                        }
                    }
                }
            }
            result.add(sb.toString());
            i = r;
        }
        return result;
    }

}
